### ISSUE TITLE
Fisher_Jenks pure python implementation is too slow

### ISSUE THREAD

**Issue by mlyons-tcc** [pysal/pysal#828]
It looks like a pure python implementation of Fisher_Jenks was intended, but pure Python implementations are simply too slow. My computer grinded to a halt when trying to run it over a dataset of 41k points with k=9. This should be updated with Cython, a Fortran/C implementation, or use numpy instead of the slow for loops.

- Platform information:
('posix', 'darwin')
('Darwin', 'machine.local', '15.4.0', 'Darwin Kernel Version 15.4.0: Fri Feb 26 22:08:05 PST 2016; root:xnu-3248.40.184~3/RELEASE_X86_64', 'x86_64')

- Python version: 
2.7.11

- SciPy version:
0.17.0

- NumPy version:
1.10.4

**Comment by jlaura** [comment]
I would suggest that you look at the [pPySAL parallel implementation of Fisher Jenks](https://github.com/pysal/pPysal/blob/master/fisher_jenks/fj_refactored.py) though I would caution that a 41,000 x 41,000 distance matrix and 41,000 x 9 error matrix are going to be quite RAM intensive.  This is likely why your machine ground to a halt - your memory was thrashing.

I would also consider using a sampling-based approach.  41k observations should be sufficiently large that you can get very accurate results with high performance.

If neither of those work for you, PRs are also welcome.

**Comment by mlyons-tcc** [comment]
I'm not that familiar with Jenks natural breaks algorithm, but I only see two matrices of pts+1 x classes+1 or 41,001 x 10. [_fisher_jenks_means](https://github.com/pysal/pysal/blob/master/pysal/esda/mapclassify.py#L316)

I'll check out the parallel implementation, and I'll look into this implementation more to see if I'll be able to update it with SIMD via numpy.

**Comment by mlyons-tcc** [comment]
It looks like someone already converted it to Cython: https://github.com/perrygeo/jenks

**Comment by jlaura** [comment]
The parallel version is fully vectorized as well (SIMD). I would look at a
sampled version first though, unless you really need to optimally classify
41,000 observations.

**Comment by mlyons-tcc** [comment]
First thing's first: improving the performance of the implementation of this algorithm. The poor performance is irrespective of my sample size, and due to a poor implementation in a language not intended to iterate over datasets. Sampling can simply be added before any implementation of the Fisher Jenks, but there's no reason to call into a slow implementation of an algorithm.

Bad news for pPySAL is that its performance comes up short of the Cython implementation. I got on the order of 250 msec for the 4k pts k=5 test. That's significantly faster than the implementation that I filed an Issue on, but it's still 5 times slower than the Cython implementation linked. Also, since pPySAL holds the entire distance matrix in memory, I taxed all my computing cores for some time at 40k points before crashing with out of memory. It's not memory efficient as you originally eluded.
